src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,231 @@
+     * By re-registering the watch before querying the state, we will not miss out on the latest leader change.
+     */
+    def longPollLeaderChange(retries: Int = 0): Unit = Future {
+      try {
+        if (latch.getState == LeaderLatch.State.STARTED)
+          client.getChildren
+            .usingWatcher(new CuratorWatcher {
+              override def process(event: WatchedEvent): Unit =
+                if (!_isCancelled) longPollLeaderChange()
+            })
+            .forPath(curatorLeaderLatchPath)
+        emitLeader()
+      } catch {
+        case ex: KeeperException.NoNodeException if retries < 100 =>
+          // Wait for node to be created
+          logger.info("retrying")
+          Thread.sleep(retries * 10L)
+          longPollLeaderChange(retries + 1)
+        case ex: Throwable =>
+          sq.fail(ex)
+      }
+    }(singleThreadEC)
+
+    /**
+      * Emit current leader. Does not fail on connection error
+      */
+    private def emitLeader(): Unit = {
+      val currentLeader = leaderHostPortMetric.blocking {
+        if (client.getState == CuratorFrameworkState.STOPPED) {
+          None
+        } else {
+          try {
+            Some(latch.getLeader.getId)
+          } catch {
+            case ex: Throwable =>
+              logger.error("Error while getting current leader", ex)
+              None
+          }
+        }
+      }
+
+      currentLeader match {
+        case Some(`hostPort`) =>
+          sq.offer(LeadershipState.ElectedAsLeader)
+        case otherwise =>
+          sq.offer(LeadershipState.Standby(otherwise))
+      }
+    }
+
+    private val closeHook: () => Unit = cancel
+
+    def start(): Unit = synchronized {
+      assert(!isStarted, "already started")
+      isStarted = true
+      // We register the beforeClose hook to ensure that we have an opportunity to remove the latch entry before we lose
+      // our connection to Zookeeper
+      clientCloseable.beforeClose(closeHook)
+      try {
+        logger.info("starting leader latch")
+        latch.start()
+        longPollLeaderChange()
+
+      } catch {
+        case ex: Throwable =>
+          logger.error("Error starting curator election event emitter")
+          sq.fail(ex)
+      }
+    }
+
+    override def isCancelled: Boolean = _isCancelled
+
+    override def cancel(): Boolean = synchronized {
+      require(isStarted, "not started")
+      if (!_isCancelled) {
+        clientCloseable.removeBeforeClose(closeHook)
+        _isCancelled = true
+        // shutdown hook remove will throw if already shutting down; swallow the exception and continue.
+
+        try {
+          logger.info("Closing leader latch")
+          latch.close()
+          logger.info("Leader latch closed")
+        } catch {
+          case ex: Throwable =>
+            logger.error("Error closing CuratorElectionStream latch", ex)
+        }
+        Try(sq.complete()) // if not already completed
+      }
+      _isCancelled
+    }
+  }
```

Comment:
Overall I don't think I understand how we use curator's `LeaderLatch` here. The canonical recipe seems to be very simple: ```leaderLatch.start(), await() // the leadership and close() //when not needed```
but I fail to recognize it here. Maybe we can chat and you could walk me through your thought process?

Reply:
Discussed. Added explanation to longPollLeaderChange() method. :)